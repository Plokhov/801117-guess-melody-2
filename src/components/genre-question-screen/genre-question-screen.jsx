import React from "react";
import PropTypes from "prop-types";

import {AudioPlayer} from "../audio-player/audio-player.jsx";

export class GenreQuestionScreen extends React.PureComponent {
  constructor(props) {
    super(props);

    const {question} = this.props;
    const {answers} = question;

    this.state = {
      activePlayer: -1,
      userAnswer: new Array(answers.length).fill(false),
    };
  }

  render() {
    const {
      question,
      onAnswer,
      screenIndex
    } = this.props;

    const {
      answers,
      genre,
    } = question;

    return <section className="game__screen">
      <h2 className="game__title">
        Выберите {genre} треки
      </h2>
      <form
        className="game__tracks"
        onSubmit={(evt) => {
          evt.preventDefault();
          onAnswer(this.state.userAnswer);
        }}
      >
        {answers.map((it, i) => {
          return (
            <div
              className="track"
              key={`${screenIndex} - answer-${it.id}`}
            >
              <div className="game__track">
                <AudioPlayer
                  src={it.src}
                  isPlaying={i === this.state.activePlayer}
                  onPlayButtonClick={() => this.setState({
                    activePlayer: this.state.activePlayer === i ? -1 : i
                  })}
                />
              </div>
              <div className="game__answer">
                <input
                  className="game__input visually-hidden"
                  type="checkbox"
                  name="answer"
                  value={`answer-${it.id}`}
                  id={`answer-${it.id}`}
                  checked={this.state.userAnswer[i]}
                  onChange={() => {
                    const userAnswer = [...this.state.userAnswer];
                    userAnswer[i] = !userAnswer[i];
                    this.setState({userAnswer});
                  }}
                />
                <label className="game__check" htmlFor={`answer-${it.id}`}>
                  Отметить
                </label>
              </div>
            </div>
          );
        })}
        <button
          className="game__submit button"
          type="submit"
          onClick={() => {
            this.setState({
              activePlayer: -1,
              userAnswer: new Array(answers.length).fill(false),
            });
          }}
        >
          Ответить
        </button>
      </form>
    </section>;
  }
}

GenreQuestionScreen.propTypes = {
  question: PropTypes.shape({
    type: PropTypes.oneOf([`genre`, `artist`]).isRequired,
    genre: PropTypes.oneOf([`jazz`, `blues`, `pop`, `rock`, `trance`, `metall`]),
    answers: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string.isRequired,
      src: PropTypes.string.isRequired,
      genre: PropTypes.oneOf([`jazz`, `blues`, `pop`, `rock`, `trance`, `metall`])
    })).isRequired,
  }).isRequired,
  screenIndex: PropTypes.number.isRequired,
  onAnswer: PropTypes.func.isRequired,
};
