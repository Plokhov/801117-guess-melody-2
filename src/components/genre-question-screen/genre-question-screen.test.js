import React from 'react';
import renderer from 'react-test-renderer';
import {GenreQuestionScreen} from './genre-question-screen.jsx';

it(`GenreQuestionScreen correctly renders after relaunch`, () => {
  const tree = renderer
    .create(<GenreQuestionScreen
      screenIndex={0}
      question={{
        type: `genre`,
        genre: `pop`,
        answers: [
          {
            id: `1`,
            src: `a`,
            genre: `pop`,
          },
          {
            id: `2`,
            src: `c`,
            genre: `pop`,
          },
          {
            id: `3`,
            src: `b`,
            genre: `pop`,
          },
          {
            id: `4`,
            src: `a`,
            genre: `pop`,
          },
        ],
      }}
      onAnswer={jest.fn()}
    />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
