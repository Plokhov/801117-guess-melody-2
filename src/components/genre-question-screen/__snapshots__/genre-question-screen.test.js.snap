// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreQuestionScreen is rendered correctly 1`] = `
<section
  className="game__screen"
>
  <h2
    className="game__title"
  >
    Выберите 
    rock
     треки
  </h2>
  <form
    className="game__tracks"
    onSubmit={[Function]}
  >
    <div
      className="track"
    >
      <button
        className="track__button track__button--play"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
      <div
        className="game__answer"
      >
        <input
          className="game__input visually-hidden"
          id="answer-undefined"
          name="answer"
          onChange={[Function]}
          type="checkbox"
          value="answer-undefined"
        />
        <label
          className="game__check"
          htmlFor="answer-undefined"
        >
          Отметить
        </label>
      </div>
    </div>
    <div
      className="track"
    >
      <button
        className="track__button track__button--play"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
      <div
        className="game__answer"
      >
        <input
          className="game__input visually-hidden"
          id="answer-undefined"
          name="answer"
          onChange={[Function]}
          type="checkbox"
          value="answer-undefined"
        />
        <label
          className="game__check"
          htmlFor="answer-undefined"
        >
          Отметить
        </label>
      </div>
    </div>
    <div
      className="track"
    >
      <button
        className="track__button track__button--play"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
      <div
        className="game__answer"
      >
        <input
          className="game__input visually-hidden"
          id="answer-undefined"
          name="answer"
          onChange={[Function]}
          type="checkbox"
          value="answer-undefined"
        />
        <label
          className="game__check"
          htmlFor="answer-undefined"
        >
          Отметить
        </label>
      </div>
    </div>
    <div
      className="track"
    >
      <button
        className="track__button track__button--play"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
      <div
        className="game__answer"
      >
        <input
          className="game__input visually-hidden"
          id="answer-undefined"
          name="answer"
          onChange={[Function]}
          type="checkbox"
          value="answer-undefined"
        />
        <label
          className="game__check"
          htmlFor="answer-undefined"
        >
          Отметить
        </label>
      </div>
    </div>
    <button
      className="game__submit button"
      type="submit"
    >
      Ответить
    </button>
  </form>
</section>
`;
